boards/BMS/Src/App/states/App_ChargeState.c

```diff
@@ -25,18 +34,42 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
     struct BmsWorld *world = App_SharedStateMachine_GetWorld(state_machine);
     struct BmsCanTxInterface *can_tx  = App_BmsWorld_GetCanTx(world);
     struct Charger *          charger = App_BmsWorld_GetCharger(world);
+    struct Clock *            clock   = App_BmsWorld_GetClock(world);
 
     if (!App_Charger_IsConnected(charger))
     {
         App_CanTx_SetPeriodicSignal_CHARGER_DISCONNECTED_IN_CHARGE_STATE(
             can_tx, true);
         App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
     }
+
+    if (App_Charger_HasFaulted(charger))
+    {
+        App_CanTx_SetPeriodicSignal_CHARGER_FAULT_DETECTED(can_tx, true);
+        App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+    }
+
+    // Track the time since accumulator charging started
+    const uint32_t elapsed_time_s =
+        App_SharedClock_GetCurrentTimeInSeconds(clock) -
+        App_SharedClock_GetPreviousTimeInSeconds(clock);
+
+    // Go to the fault state when charging has completed to indicate that
+    // the charger should be disconnected and the BMS-Core should be
+    // power-cycled
+    if (elapsed_time_s >= TOTAL_CHARGE_TIME_S)
+    {
+        App_CanTx_SetPeriodicSignal_CHARGING_DONE(can_tx, true);
+        App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
```

Comment:
I think we are supposed to transition to the init state

Reply:
Might be a good idea to allow transition to the init state when we have completed charging AND physically disconnect the charger to avoid the possibility of cycling back into the charge state

Reply:
> Has this been tested with the charger @timbitss ?

no we plan to asap once joe and edwin get the wiring resolved

Reply:
> Might be a good idea to allow the transition to the init state when we have completed charging AND physically disconnect the charger to avoid the possibility of cycling back into the charge state

Yes, I think that definitely makes the most sense. I, therefore, believe we need to keep the "charging complete" CAN signal to indicate to the user they should disconnect the charger. What do you think about adding another CAN signal with the status of the PON pin? I don't think it hurts IMO, helps for debugging

Reply:
Sure yeah having PON would be good for debugging.

Reply:
If we are to follow the rules for charging. We will need to charge on the cart. This means that in order to put the accumulator back in the car we will need to power cycle the BMS anyways (power comes from the charger lemo). I feel that even when charging on the car, a power cycle would be a good idea. It's an easy way to avoid further accidental precharge scenarios. For example, say we finish charging and enter into the init state, if the shutdown loop is closed again, we could re-enter the charging state accidentally. I think we should try to avoid these accidental precharges when changing connectors to the accumulator so a power off would be the best option.